src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivitySource.cs

```diff
@@ -407,105 +407,103 @@ internal void NotifyActivityAddException(Activity activity, Exception exception,
         }
     }
 
-    // SynchronizedList<T> is a helper collection which ensure thread safety on the collection
-    // and allow enumerating the collection items and execute some action on the enumerated item and can detect any change in the collection
-    // during the enumeration which force restarting the enumeration again.
-    // Caution: We can have the action executed on the same item more than once which is ok in our scenarios.
+    //this class uses Interlocked operations and a copy-on-write design to ensure thread safety
+    //all operations are thread safe
     internal sealed class SynchronizedList<T>
     {
-        private readonly List<T> _list;
-        private uint _version;
-
-        public SynchronizedList() => _list = new List<T>();
+        //This array must not be written to directly. Copy the array and then replace it with the new array.
+        private T[] _volatileArray;
+        public SynchronizedList() => _volatileArray = [];
 
         public void Add(T item)
         {
-            lock (_list)
+            while (true)
             {
-                _list.Add(item);
-                _version++;
+                T[] local = _volatileArray;
+
+                if (TryAppendItemAndSwap(item, local))
+                {
+                    return;
+                }
```

Comment:
why can't simplify this by just doing something like:

```C#
lock (this)
{
  ... update the _volatileArray
}
```

mutating the list should be rare and even if i t happen will be done at different time which will not cause lock contention.

Reply:
2 questions
- Are you suggesting to use a `List<T>` again?
  - if so doesn't that introduce instability to worker/request threads accesing the list that might be mutating?
-  or you are suggesting continuing to use immutable arrays, but only lock the mutating operations and leave the reader/request threads unlocked?

Reply:
I meant the later `or you are suggesting continuing to use immutable arrays, but only lock the mutating operations and leave the reader/request threads unlocked?`

Reply:
That should be fine, I can make these changes and re-test

Reply:
Thanks a lot for helping with the fix.

Reply:
Fixed

Reply:
This is what I've suggested earlier. Should've been more clear with intent I guess :)

Reply:
> This is what I've suggested earlier. Should've been more clear with intent I guess :)

Ah, gotcha. I knew this was possible, i just didn't know if the weaker semantics were tolerable by the design.